### PR TITLE
(refactor) purchase orders api

### LINKED ISSUE #480: Proposal: Remove (specific) support for donors
In 1.x we has a specific table for `donor`s in the database.  I propose that we model donors as suppliers in 2.x and remove support for registering donors.  If we need to, we can put a flag in the supplier table (`is_donor`).

@DedrickEnc, @mbayopanda based on your experience with 1.x's donation management, is this possible?

### LINKED ISSUE #512: (Design) Purchase Orders
This issue proposes the database schema for purchase orders in 2.x

**Schema**
The purchase table has a lot information unrelated to the actual purchase order to be sent to the supplier.   We should start small with only the necessary information and try to add in more if we need it.

| column | type | description |
| --- | --- | --- |
| uuid | BINARY |  |
| project_id | INT |  |
| reference | INT | A generated ID that, combined with `project_id`, creates a human readable id |
| cost | DECIMAL | the total cost of the purchase order |
| date | DATETIME | the date of the purchase order |
| currency_id | INT | the id of the currency the order is made in |
| note | TEXT | some text describing the nature of the purchase order |
| creditor_uuid | BINARY | the creditor to be sent the purchase order |
| created_at | TIMESTAMP | set at the server time when the record was created |

The table `purchase_item` doesn't need to change.

For the first iteration, this will be all the properties stored in the purchase order table.

**HTTP API**
The HTTP API will follow the standard [API Guidelines](https://github.com/IMA-WorldHealth/bhima-2.X/wiki/Design:-Server-CRUD-APIs).  You will not be able to delete a Purchase Order yet, until that case is analyzed.

### PR DESCRIPTION
**Pull request by jniles**
This PR rewrites the purchasing backend tables and API.

The design from #512 is implemented here.  In order to prepare for other inventory, stock, and purchasing modules, this PR also cleans out unused, broken, or old APIs on the server.

**Major Changes**
1. Removes support for `donors` and `donations`.  If these are necessary, they can be modeled in the existing suppliers and inventory infrastructure.  See the discussion on #480.
2. Removes `categorised/`.  After re-classifying a few files as finance or admin, there were no more files in `categorised/` and it has been removed.
3. Removes `/data` routes, `data.js`, `parser.js`.
4. Removes all API and end to end test associated with `donors`.  Rewrites the `purchase` API tests to meet the [API Guidelines](https://github.com/IMA-WorldHealth/bhima-2.X/wiki/Testing:-Integration-Testing-Server-APIs).

Closes #512.  Closes #480.

---

Thank you for contributing!

Before submitting this pull request, please verify that you have:
- [x] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
- [x] Run integration tests.
- [x] Run end-to-end tests.
- [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

### CODE REVIEW

**Comment by jniles** [comment]
@sfount, could I get a review?

**Comment by jniles** [comment]
Ping @sfount.

File: server/controllers/finance/purchases.js [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+ * @description
+ * POST /purchases
+ *
+ * Creates a purchase order in the database
+ */
+function create(req, res, next) {
+  let data = req.body;
+
+  if (!data.items) {
+    return next(
```
Comment (sfount) [inline server/controllers/finance/purchases.js:130]: I've noticed that in some places you are throwing Error objects and in others passing them through to `next`.

In methods that do not have access to `next` we would have to use throw but it seems to be that `throw` here would pass the object through to `next` in the same way.

Do you have a preference?

**Comment by sfount** [comment]
This LGTM - excellent clean-up of legacy code too!

File: server/controllers/finance/purchases.js (hunk 1 above)
Comment (jniles) [inline server/controllers/finance/purchases.js:130]: `throw` is only used inside promise chains - using `throw` here would actually raise an uncaught exception.

In general, what I've been doing is `if in promise chain; throw; else return next()`.   I prefer `throw`ing myself, but it isn't always practical to have `next()` available, as you pointed out.

**Comment by jniles** [comment]
@sfount, rebased with master.  If tests pass, this could be merged anytime.

**Pull request merged**